lib/smart_proxy_dynflow_core/log.rb

```diff
@@ -1,61 +1,112 @@
-require 'logger'
+require 'logging'
 
 module SmartProxyDynflowCore
-  class Log < ::Logger
-    alias_method :write, :debug
+  class Log
+    BASE_LOG_SIZE = 1024 * 1024 # 1 MiB
+    LOGGER_NAME = 'dynflow-core'.freeze
 
-    class << self
-      def instance
-        if @logger.nil?
-          @logger = self.new log_file
-          @logger.level = log_level
-        end
-        @logger
-      end
-
-      def instance=(logger)
-        @logger = logger
-      end
+    begin
+      require 'syslog/logger'
+      @syslog_available = true
+    rescue LoadError
+      @syslog_available = false
+    end
+    @reopen = false
 
+    class << self
       def reload!
+        Logging.logger[LOGGER_NAME].appenders.each(&:close)
+        Logging.logger[LOGGER_NAME].clear_appenders
         @logger = nil
         instance
       end
 
-      def log_level
-        if Settings.instance.loaded && Settings.instance.log_level
-          ::Logger.const_get(Settings.instance.log_level.upcase)
+      def reopen
+        return if @logger.nil?
+        if Settings.instance.log_file !~ /^(STDOUT|SYSLOG|JOURNALD?)$/i
+          @logger.warn "Scheduled log file reopen"
+          @reopen = true
         else
-          Logger::WARN
+          @logger.debug "Not scheduling log file reopen (no file appender)"
         end
       end
 
-      def log_file
-        if Settings.instance.loaded && Settings.instance.log_file
-          Settings.instance.log_file
+      def handle_reopen
+        @logger.warn "Performing log file reopen"
+        @reopen = false
+        @logger.appenders.each(&:reopen)
```

Comment:
There still seems to be some flaw there: when I try:

1. `mv logs/smart_proxy_dynflow_core.log{,.bak}`
2. `pkill -f -USR1 bin/smart_proxy_dynflow_core`
3. `curl localhost:8008/tasks/ -X POST -d '{ "action_name": "SmartProxyDynflowCore::TestAction", "action_input": {"exception": "false"} }'`
4.  `du logs/smart_proxy_dynflow_core.log`:
```
0	logs/smart_proxy_dynflow_core.log
```

So the new file is re-created, but no output is getting in. Seems like there is some additional logic in smart-proxy that actually makes it work (the smart-proxy worked fine when trying this)

Reply:
Good find, yeah the issue is that I call `handle_reopen` only in `instance` which is obviously only called during initialization. I wanted to avoid overloading logging methods but it looks like it's the only way.